### PR TITLE
move update logic into `started` since we should only care about update

### PR DESCRIPTION
**Pull request by acornett21**
- Fixes: #596 
- Moving `updateProject` logic into the `Started` code since we only need to care/worry about updating this status on net new projects
- Added additional comments for uses cases when/how the project status can move
- Added logic to convert `index.docker.io` to `docker.io` since more projects use this value today, and probably is the more preferred approach 
- Added Unit tests to to cover new logic
- Updated mock so GET vs PATCH for Project route both can get called in the submit flow

Code coverage is closer to what it was before, but still not where it needs to be. But these change now fully cover the `Started` logic.
image

Signed-off-by: Adam D. Cornett <adc@redhat.com>

### CODE REVIEW

File: certification/pyxis/submit_test.go [hunk 1]
```diff
@@ -49,7 +49,83 @@ var _ = Describe("Pyxis Submit", func() {
...
+									Registry:   "my.registry",
+									Repository: "my/repo",
+								},
+							},
+						},
+						RpmManifest: &RPMManifest{},
+						TestResults: &TestResults{},
+						Artifacts:   []Artifact{},
+					})
+					Expect(err).To(MatchError(errors.New("error calling remote API")))
```
Comment (komish) [inline certification/pyxis/submit_test.go:121]: This strikes me as a brittle comparison, potentially because it's too granular?

File: certification/pyxis/submit_test.go [hunk 2]
```diff
@@ -49,7 +49,83 @@ var _ = Describe("Pyxis Submit", func() {
 		})
 	})
 
-	Context("updateProject 401 Unauthorized", func() {
+	Context("when a project is submitted with an empty registry", func() {
+		BeforeEach(func() {
```
[the remark this answers is not in the document]
Comment (acornett21) [reply certification/pyxis/submit_test.go:53]: I can't speak to best practices for Ginko, every go project I've worked on just does whatever it wants, with no rhyme or reason, our project is no different with all tests in this package being structured this way. I don't see that Ginko has a setup or a before? Maybe if that existed it would be more valid for our tests.

**Review by komish — COMMENTED** [review]
couple of things

File: certification/pyxis/submit_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (komish) [reply certification/pyxis/submit_test.go:53]: So this was done to be consistent with the rest of the test file?

File: certification/pyxis/submit.go [hunk 3]
```diff
@@ -17,31 +20,42 @@ func (p *pyxisClient) SubmitResults(ctx context.Context, certInput *certificatio
...
 
-	// Set this project's metadata to match the image that we're certifying.
-	certProject.Container.Registry = certImage.Repositories[0].Registry
-	certProject.Container.Repository = certImage.Repositories[0].Repository
+		// Setting registry to the value we get from certImage from crane and then normalizing
+		// index.docker.io to docker.io so project info shows properly in the Red Hat Catalog
+		registry := certImage.Repositories[0].Registry
+		if registry == name.DefaultRegistry {
+			registry = defaultRegistryAlias
+		}
```
[the remark this answers is not in the document]
Comment (acornett21) [reply certification/pyxis/submit.go:41]: We never update the project after the first image is submitted. Since this value cannot be changed on the project page in the UI, we need to normalize that upfront. The value in the UI at the project level drives what ends up in the catalog and for people to `pull` the image via this value. So it only makes sense to normalize here and not everwhere in our code since we still ned `index.docker.io` in the `crane` values so we can still do authenticated calls for say `list-tags` or anything else that would require auth.

TL:DR only net new cert projects are updated by preflight.

[the remark this answers is not in the document]
Comment (komish) [reply certification/pyxis/submit.go:41]: I'm not following; I'll let someone else review here.

[the remark this answers is not in the document]
Comment (komish) [reply certification/pyxis/submit.go:41]: Spoke with @acornett21.

This bit changes the registry listing for effectively new projects to `docker.io` if crane internally parsed the image's registry to live in Docker Hub. 

Part of the issue for me is that this value doesn't get set to what the user provided. It's definitely coming from Docker hub, so in practice, it's mostly no harm and no foul. It just seemed strange.

Secondly, this doesn't change the image registry value to what the user provided when submitting to pyxis. Again, I would say it's somewhat fair to assume that the value we give pyxis (which I'm fairly certain is always index.docker.io) should function as expected, and so changing the image value back to what the user gave us may not net us benefit. It just feels inconsistent.

Either way, functionally, I see where this is going. Seems like it should work.

I'll leave this unresolved for folks to review to see if it seemed odd or if it was just me (entirely possible, as I don't know if I understand the submit flow that well).

**Review by komish — APPROVED** [review]
/lgtm

**Review by bcrochet — APPROVED** [review]
Just a comment. Otherwise,

/lgtm

File: certification/pyxis/submit.go [hunk 4]
```diff
@@ -17,31 +20,42 @@ func (p *pyxisClient) SubmitResults(ctx context.Context, certInput *certificatio
 	certImage := certInput.CertImage
 
 	// Submission effectively starts the certification process, so switch
-	// the status to reflect this if needed.
+	// the status to reflect this if needed. This only needs to be done for net new projects.
+	// Existing projects that are in "In Progress" can stay "In Progress" until they moved to "Published" which is triggered
+	// once an image in a project is moved to "Published" status. The status on the project would stay in "Published" status,
+	// unless the partner decides to un-publish all of their images. At that point backed systems/processes would move
+	// the project back to "In Process" and there would still be nothing that preflight need to update on the project.
 	if certProject.CertificationStatus == "Started" {
```
[the remark this answers is not in the document]
Comment (acornett21) [reply certification/pyxis/submit.go:28]: Yeah, I was wondering that as well, that's why I asked the CVP/Pyxis team. These are the scenarios they covered, so if there is another one that exists that they don't know about, that's kind of problematic, so hopefully they explained all of them and I didn't lose any in translation. I'm not sure if this would help, but maybe we need to have a meeting with them to explain all the backend systems and their role in this process.
